// hdma-geoviewer/nodejs/config.js
module.exports={
  server:{
    port:8080,
    api:false //have not finished yet.
  },
  twitter:{
    consumer_key: '', // can get it from https://apps.twitter.com/
    consumer_secret: '',// can get it from https://apps.twitter.com/
    access_token: '',// can get it from https://apps.twitter.com/
    access_token_secret: ''// can get it from https://apps.twitter.com/
  },
  mongodb:{
    ip:"127.0.0.1",
    port:"27017",
    username:null, //no username at this time.
    password:null, //no password at this time.
    db:"HDMA", //database name
    collection:"twitterStreamingDB" //the collection you want to save tweets
  },
  areas:[
    {
      startDate:"2014-12-05T16:00:00.000-0800", //iso date format
      endDate:"2015-12-05T16:00:00.000-0800",//iso date format
      area:"sanDiego", //the area you want to name it. It will save in the "search_region" field under the collection in the mongodb
      keyword:null, // can be "twitter api, twitter streaming" for multiple keywords
      bbox: "-117.595898,32.534180,-115.835742,33.507161", // twitter streaming bbox
      filterParameter: "location",
      host:"both", //please not change this at this time. may need to change in the future.
      blacklists:["Longdickya9","vincenvi619","vme7413", "uruguaydiablo", "just_sanch","bhaluPnMC","Sunnyrajkumar1","MoneyOrtega","lilcthebakedgod","BixberNavy","genitalia666","EarneyHarisson", "juanesp64250381", "skylerhavenxxx", "debmootoo"] //twitter usernames
    }
  ],
  googleAnalytics:"",
  gmail:{
    email:"FROM_GMAIL",
    emailTo:"to email1, to email2",
		clientID:"", //from google developer console OAuth 2.0 client IDs
		clientSecret:"", //from google developer console OAuth 2.0 client IDs
		refreshToken:"", //please refer https://www.youtube.com/watch?v=hfWe1gPCnzc to see how to get refersh token and access token by using oauth playground
		accessToken:"" //please refer https://www.youtube.com/watch?v=hfWe1gPCnzc to see how to get refersh token and access token by using oauth playground
  }
}
